proposals/migrate-over-ipfs.md

```diff
@@ -0,0 +1,163 @@
+# Download migrations over IPFS
+
+Authors: Stebalien
+
+Initial PR: TBD <!-- Reference the PR first proposing this document. Oooh, self-reference! -->
+
+## Purpose &amp; impact 
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+
+Currently, when the on-disk layout of the datastore changes, go-ipfs will download a "migration"
+tool over HTTPS as-needed to "upgrade" the datastore. Unfortunately, this happens over HTTPS so:
+
+1. We're not dogfooding our own tech.
+2. It doesn't work in places where https://dist.ipfs.io is blocked (e.g., China).
+
+At the moment, we're unable to ask Brave to update go-ipfs to the latest version because the latest
+version will need to download one of these migrations and many of Brave's users are in China.
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+
+- IPFS in Brave needs to matter.
+- A significant number of Brave users need to be in China.
+- Fetching migrations over IPFS needs to be reliable.
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+
+- When starting the go-ipfs daemon with the `--migrate` flag, the migrations would be fetched over IPFS instead of HTTPs.
+- When starting the go-ipfs daemon in a bundled application like IPFS Desktop or Brave, the user shouldn't notice anything except that we make no connections to https://ipfs.io when starting go-ipfs the first time after upgrading.
+
+#### Impact
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+- This would unblock updating go-ipfs in Brave, allowing us to ship bug fixes and new features to Brave users.
+- This would prove out IPFS as a way to ship code/updates to users.
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+This is ones step towards our goal of updating go-ipfs itself over IPFS. In terms of knowledge, not much.
+
+#### Confidence
+_How sure are we that this is the right problem to tackle? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+3?
+
+- It's a frequent source of upgrade problems: https://github.com/ipfs/ipfs-desktop/issues?q=is%3Aissue+fs-repo-migrations
+- It's preventing us from shipping new go-ipfs releases (with important fixes) to brave users.
+- It hurts adoption in China.
+
+## Project definition
+
+#### Brief plan of attack
+
+**Design sketch:**
+
+Given a config section like:
+
+```js
+{
+    "Migration": {
+        // Sources in order of preference where "HTTPS" means our gateways and "IPFS" means over IPFS.
+        // (an empty list means "do the default thing")
+        "DownloadPolicy": ["HTTPS", "custom-gateway.io", "IPFS"],
+        // Whether or not to keep the migration after downloading it.
+        "Keep": "pin"|"cache"|"discard"
+    }
+}
```

Comment:
Should this be read from the existing config file, or something that a user puts in a separate migration config file?  The existing config file may not be readable until after migration is performed.

What about having this specified by an optional CLI flag?
```
--dlpolicy="https,ipfs,custom-gateway.io,..." 
```

Reply:
GAHHHHHHH. This is so, very, annoying.

What if we _just_ decoded this part of the config? As long as this part of the config remains somewhat consistent across upgrades (and we correctly handle missing fields), we should be fine.

Basically:

1. Get the path to the repo.
2. Read the config file into a `struct { Migrations config.MigrationConfig }` type.
3. Move on.

This is significantly safer than, say, reading the swarm config, because it _only_ deals with migrations.

Thoughts?

Reply:
That is fine. Will go with config file.

Reply:
I think it should also include:
```go
    // Peers lists the nodes to attempt to connect with when downloading migrations.                                                                     
    Peers []peer.AddrInfo                                                                                                           
```